### ISSUE TITLE
Exceptions during Kubernetes pod shutdown

### ISSUE THREAD

**Issue by kdrach** [dotnet/orleans#8410]
We are experiencing exceptions during pod shutdown on a Kubernetes cluster. We have set up SQL clustering with Kubernetes hosting using the following configuration:

```            silo.AddAdoNetGrainStorageAsDefault(opt =>
            {
                opt.ConnectionString = databaseConnection;
                opt.Invariant = InvariantNameSqlServer;
            });

            silo.UseAdoNetClustering(opt =>
            {
                opt.ConnectionString = databaseConnection;
                opt.Invariant = InvariantNameSqlServer;
            });
            silo.UseKubernetesHosting();
```
We have also set up SQL health checks and a general health check to validate if the silos are in an active state.
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: "{{ .Chart.Name }}-deployment"
  labels:
    app: {{ .Chart.Name }}
    orleans/serviceId: {{ .Chart.Name }}
spec:
  selector:
    matchLabels:
      app: {{ .Chart.Name }}
      orleans/serviceId: {{ .Chart.Name }}
  template:
    metadata:
      labels:
        app: {{ .Chart.Name }}
        orleans/serviceId: {{ .Chart.Name }}
        orleans/clusterId : {{ .Chart.Name }}
        .
        .
        .
        containers:
        - name: {{ .Chart.Name }}
        .
        .
        .
          env:
            - name: ORLEANS_SERVICE_ID
              valueFrom:
                fieldRef:
                  fieldPath: metadata.labels['orleans/serviceId']
            - name: ORLEANS_CLUSTER_ID
              valueFrom:
                fieldRef:
                  fieldPath: metadata.labels['orleans/clusterId']
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: POD_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          envFrom:
            - configMapRef:
                name: env-file
          ports:
            - containerPort: 80
          livenessProbe:
            httpGet:
              path: /health/liveness
              port: 80
            initialDelaySeconds: 5
            periodSeconds: 2
            timeoutSeconds: 2
          readinessProbe:
            httpGet:
              path: /health/readiness
              port: 80
            initialDelaySeconds: 5
            periodSeconds: 3
            timeoutSeconds: 2
```
Stack traces:

1st exception: 
```
 Orleans.Runtime.OrleansMessageRejectionException: Exception while sending message: 
 Orleans.Runtime.Messaging.ConnectionFailedException: Unable to connect to endpoint S. See InnerException  ---> 
 Microsoft.AspNetCore.Connections.ConnectionAbortedException: Underlying connection closed    at 
 Orleans.Internal.OrleansTaskExtentions.MakeCancellable(Task task, CancellationToken cancellationToken) in 
 /_/src/Orleans.Core/Async/TaskExtensions.cs:line 190    at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress 
 address, ConnectionEntry entry) in /_/src/Orleans.Core/Networking/ConnectionManager.cs:line 228    --- End of inner exception stack 
 trace ---    at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address, ConnectionEntry entry) in 
 /_/src/Orleans.Core/Networking/ConnectionManager.cs:line 228    at 
 Orleans.Runtime.Messaging.ConnectionManager.GetConnectionAsync(SiloAddress endpoint) in 
 /_/src/Orleans.Core/Networking/ConnectionManager.cs:line 108    at Orleans.Runtime.Messaging.MessageCenter. <SendMessage>g__SendAsync|30_0(MessageCenter messageCenter, ValueTask`1 connectionTask, Message msg) in 
 /_/src/Orleans.Runtime/Messaging/MessageCenter.cs:line 231    at 
 Orleans.Serialization.Invocation.ResponseCompletionSource.GetResult(Int16 token) in 
 /_/src/Orleans.Serialization/Invocation/ResponseCompletionSource.cs:line 90    at Orleans.Runtime.OutgoingCallInvoker`1.Invoke() in 
 /_/src/Orleans.Core/Runtime/OutgoingCallInvoker.cs:line 129    at 
 Orleans.Runtime.ActivityPropagationGrainCallFilter.Process(IGrainCallContext context, Activity activity) in 
 /_/src/Orleans.Core/Diagnostics/ActivityPropagationGrainCallFilter.cs:line 75    at Orleans.Runtime.OutgoingCallInvoker`1.Invoke() in 
 /_/src/Orleans.Core/Runtime/OutgoingCallInvoker.cs:line 129    at 
 Orleans.Runtime.GrainReferenceRuntime.InvokeMethodWithFiltersAsync(GrainReference reference, IInvokable request, 
 InvokeMethodOptions options) in /_/src/Orleans.Core/Runtime/GrainReferenceRuntime.cs:line 83    at 
 Orleans.Runtime.GrainDirectory.ClientDirectory.PublishUpdates() in /_/src/Orleans.Runtime/GrainDirectory/ClientDirectory.cs:line 499
```
second one:
```
 System.ObjectDisposedException: Cannot access a disposed object. Object name: 'This instance has been disposed'.    at  
 Orleans.Runtime.Utilities.AsyncEnumerable`1.AsyncEnumerator.System.Collections.Generic.IAsyncEnumerator<T>.MoveNextAsync() in 
 /_/src/Orleans.Core/Utils/AsyncEnumerable.cs:line 133    at 
 Orleans.Hosting.Kubernetes.KubernetesClusterAgent.MonitorOrleansClustering() in 
 /_/src/Orleans.Hosting.Kubernetes/KubernetesClusterAgent.cs:line 137    at 
 Orleans.Hosting.Kubernetes.KubernetesClusterAgent.MonitorOrleansClustering() in 
 /_/src/Orleans.Hosting.Kubernetes/KubernetesClusterAgent.cs:line 137
```

We are also experiencing that new silos require about 30-45 seconds for cleanup after the previous silos have failed.

**Comment by m3nax** [comment]
I think the problem depends on the default time that k8s guarantees before sending KILL signal to the container.
By default it waits 30 seconds ([docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)) after which it executes the killing of the container, i think orleans takes longer to do a controlled shutdown in your enviroment.

Try chage `terminationGracePeriodSeconds` to a greater value like **180**, as it is written in the Microsoft guide for clustering with k8s ([docs](https://learn.microsoft.com/en-us/dotnet/orleans/deployment/kubernetes)).

**Comment by ReubenBond** [comment]
This ought to have been fixed by #9048
